File: src/main/java/com/qwazr/jdbc/cache/DatatypeConverter.java
package com.qwazr.jdbc.cache;

import java.nio.charset.StandardCharsets;
import java.util.Formatter;

public class DatatypeConverter {
    public static String printHexBinary(byte[] data) {
        Formatter formatter = new Formatter();
        for (byte b : data) {
            formatter.format("%02x", b);
        }
        String hex = formatter.toString();
        formatter.close();
        return hex;
    }

    public static byte[] parseHexBinary(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i+1), 16));
        }
        return data;
    }

    public static String printBase64Binary(byte[] data) {
        return java.util.Base64.getEncoder().encodeToString(data);
    }

    public static byte[] parseBase64Binary(String s) {
        return java.util.Base64.getDecoder().decode(s);
    }

    public static String printHexBinary(String s) {
        return printHexBinary(s.getBytes(StandardCharsets.UTF_8));
    }

    public static String parseString(String s) {
        return new String(parseHexBinary(s), StandardCharsets.UTF_8);
    }
}

File: src/main/kotlin/io/github/jhstatewide/jdbc/cache/DatatypeConverter.kt
package io.github.jhstatewide.jdbc.cache

import java.nio.charset.StandardCharsets
import java.util.*

object DatatypeConverter {
    fun printHexBinary(data: ByteArray): String {
        val formatter = Formatter()
        for (b in data) {
            formatter.format("%02x", b)
        }
        val hex = formatter.toString()
        formatter.close()
        return hex
    }

    fun parseHexBinary(s: String): ByteArray {
        val len = s.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((s[i].digitToIntOrNull(16) ?: (-1 shl 4)) + s[i + 1].digitToIntOrNull(16)!!).toByte()
            i += 2
        }
        return data
    }

    fun printBase64Binary(data: ByteArray?): String {
        return Base64.getEncoder().encodeToString(data)
    }

    fun parseBase64Binary(s: String?): ByteArray {
        return Base64.getDecoder().decode(s)
    }

    fun printHexBinary(s: String): String {
        return printHexBinary(s.toByteArray(StandardCharsets.UTF_8))
    }

    fun parseString(s: String): String {
        return parseHexBinary(s).toString(StandardCharsets.UTF_8)
    }
}